internal/nix/search.go

```diff
@@ -106,3 +104,85 @@ func searchSystem(url, system string) (map[string]*Info, error) {
 	}
 	return parseSearchResults(out), nil
 }
+
+// searchSystemCache is a machine-wide cache of search results. It is shared by all
+// Devbox projects on the current machine. It is stored in the XDG cache directory.
+type searchSystemCache struct {
+	QueryToInfo map[string]map[string]*Info `json:"query_to_info"`
+}
+
+const (
+	// searchSystemCacheSubDir is a sub-directory of the XDG cache directory
+	searchSystemCacheSubDir   = "devbox/nix"
+	searchSystemCacheFileName = "search-system-cache.json"
+)
+
+var cache = searchSystemCache{}
+
+// SearchNixpkgsAttribute is a wrapper around searchSystem that caches results.
+// NOTE: we should be very conservative in where we use this function. `nix search`
+// accepts generalized `installable regex` as arguments but is slow. For certain
+// queries of the form `nixpkgs/<commit-hash>#attribute`, we can know for sure that
+// once `nix search` returns a valid result, it will always be the very same result.
+// Hence we can cache it locally and answer future queries fast, by not calling `nix search`.
+func SearchNixpkgsAttribute(query string) (map[string]*Info, error) {
+	if cache.QueryToInfo == nil {
+		contents, err := readSearchSystemCacheFile()
+		if err != nil {
+			return nil, err
+		}
+		cache.QueryToInfo = contents
+	}
+
+	if result := cache.QueryToInfo[query]; result != nil {
+		return result, nil
+	}
+
+	info, err := searchSystem(query, "" /*system*/)
+	if err != nil {
+		return nil, err
+	}
+
+	cache.QueryToInfo[query] = info
+	if err := writeSearchSystemCacheFile(cache.QueryToInfo); err != nil {
+		return nil, err
+	}
+
+	return info, nil
+}
+
+func readSearchSystemCacheFile() (map[string]map[string]*Info, error) {
```

Comment:
Thoughts on using something like https://github.com/jetpack-io/opensource/blob/main/envsec/internal/filecache/filecache.go#L2-L2 ? 

(it's in internal but we can pull it out)

Caching per result is a tiny bit more inefficient but has significant benefits:

* No race conditions
* Per result expiration
* Easy to inspect file system

Using that library this could look like

```golang
var searchCache := filecache.New("jetpack.io")
func CachedSearch(query string) (map[string]*Info, error) {
  if result, err := searchCache.Get(query); if err == nil {
    // unrmarshall and return
  }
  info, err := searchSystem(query, "" /*system*/)
  if err != nil {
    return nil, err
  }
  marshalledInfo, err := json.Marshall(info)
  if err != nil {
    return nil, err
  }
  if err = searchCache.set(query, marshalledInfo, time.Hour * 24 * 30); err != nil {
    return nil, err
  }
  return info, nil
}
```

Reply:
Another benefit of reusing code is to not have a custom cache solution and instead reusing code we already have. This would be the main reason.

> The "per result expiration" doesn't apply here. We should only cache for queries that always resolve to the same result.

Disagree, expiration is always good. Especially for performance only cache. If you make expiration 1 month, you know that if you ever change the code, the entire cache will turnover in a month. We also can't rely on nix search returning the same thing forever.

> Again, I don't think this should be CachedSearch for reasons explained above.

I'm not sure I understand the hesitation of naming it `CachedSearch`. It precisely what it is doing. Your above comment says:

> I wanted to keep it specific to search queries I have high confidence should always resolve to the very same result. The nix search api accepts installable regex queries which this doesn't. Hence I chose a very specific name to indicate what scenarios this works with.

But naming it `SearchNixpkgsAttribute` doesn't tell me at all that the result is cached. 

I'm not sure it matters if it doesn't support regex (does it not?) as long as we document it no?